go/src/client/lc0_main.go

```diff
@@ -0,0 +1,505 @@
+		if err != nil {
+			log.Fatal(err)
+		}
+		fmt.Printf("Cleanup training files:\n")
+		for _, f := range files {
+			fmt.Printf("%s/%s\n", train_dir, f.Name())
+		}
+		err = os.RemoveAll(train_dir)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	return nil
+}
+
+type gameInfo struct {
+	pgn   string
+	fname string
+}
+
+type CmdWrapper struct {
+	Cmd      *exec.Cmd
+	Pgn      string
+	Input    io.WriteCloser
+	BestMove chan string
+	gi       chan gameInfo
+	Version  string
+}
+
+func (c *CmdWrapper) openInput() {
+	var err error
+	c.Input, err = c.Cmd.StdinPipe()
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func convertMovesToPGN(moves []string) string {
+	game := chess.NewGame(chess.UseNotation(chess.LongAlgebraicNotation{}))
+	for _, m := range moves {
+		err := game.MoveStr(m)
+		if err != nil {
+			log.Fatalf("movstr: %v", err)
+		}
+	}
+	game2 := chess.NewGame()
+	b, err := game.MarshalText()
+	if err != nil {
+		log.Fatalf("MarshalText failed: %v", err)
+	}
+	game2.UnmarshalText(b)
+	return game2.String()
+}
+
+func CreateCmdWrapper() *CmdWrapper {
+	c := &CmdWrapper{
+		gi:       make(chan gameInfo),
+		BestMove: make(chan string),
+	}
+	return c
+}
+
+func (c *CmdWrapper) launch(networkPath string, args []string, input bool) {
+	weights := fmt.Sprintf("--weights=%s", networkPath)
+	dir, _ := os.Getwd()
+	lcargs := []string{"selfplay", weights, "--training=true", "--visits=800"}
+	lcargs = append(lcargs, lc0_args...)
+	c.Cmd = exec.Command(path.Join(dir, "lc0"), lcargs...)
+	// TODO: server args are hosed right now.
+	//	c.Cmd.Args = append(c.Cmd.Args, args...)
+	if *gpu != -1 {
+		c.Cmd.Args = append(c.Cmd.Args, fmt.Sprintf("--gpu=%v", *gpu))
+	}
+	if !*debug {
+		//		c.Cmd.Args = append(c.Cmd.Args, "--quiet")
+		fmt.Println("lc0 is never quiet.")
+	}
+	fmt.Printf("Args: %v\n", c.Cmd.Args)
+
+	stdout, err := c.Cmd.StdoutPipe()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	stderr, err := c.Cmd.StderrPipe()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	go func() {
+		defer close(c.BestMove)
+		defer close(c.gi)
+		stdoutScanner := bufio.NewScanner(stdout)
+		for stdoutScanner.Scan() {
+			line := stdoutScanner.Text()
+			//			fmt.Printf("lc0: %s\n", line)
+			switch {
+			case strings.HasPrefix(line, "gameready"):
+				parts := strings.Split(line, " ")
+				if parts[1] != "trainingfile" ||
+					parts[3] != "gameid" ||
+					parts[5] != "player1" ||
+					parts[7] != "result" ||
+					parts[9] != "moves" {
+					log.Printf("Malformed gameready: %q", line)
+					break
+				}
+				file := parts[2]
+				//gameid := parts[4]
+				//result := parts[8]
+				pgn := convertMovesToPGN(parts[10:])
+				fmt.Printf("PGN: %s\n", pgn)
+				c.gi <- gameInfo{pgn: pgn, fname: file}
+			case strings.HasPrefix(line, "bestmove "):
+				//				c.BestMove <- strings.Split(line, " ")[1]
+			case strings.HasPrefix(line, "id name lczero "):
+				c.Version = strings.Split(line, " ")[3]
+			case strings.HasPrefix(line, "info"):
+				// just swallow this line.
```

Comment:
It would be nice to show info I guess.
Because without it it's not visible that something happens, and also it would be nice to see nps.

Reply:
re-enabled